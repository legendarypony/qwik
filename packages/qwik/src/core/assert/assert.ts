import { logErrorAndStop } from '../util/log';
import { qDev } from '../util/qdev';
import { isString } from '../util/types';

export const assertDefined = (value: any, text?: string) => {
  if (qDev) {
    if (value != null) return;
    throw logErrorAndStop(text || 'Expected defined value');
  }
};

export const assertNotPromise = (value: any, text?: string) => {
  if (qDev) {
    if (!(value instanceof Promise)) return;
    throw logErrorAndStop(text || 'Expected defined value.');
  }
};

export const assertDefinedAndNotPromise = (value: any, text?: string) => {
  if (qDev) {
    assertDefined(value, text);
    assertNotPromise(value, text);
  }
};

export const assertInstanceOf = (value: any, type: any, text?: string) => {
  if (qDev) {
    if (value instanceof type) return;
    throw logErrorAndStop(
      text || `Expected value '${value}' to be instance of '${type}' but was '${typeOf(value)}'.`
    );
  }
};

export const assertString = (value: any, text?: string) => {
  if (qDev) {
    if (isString(value)) return;
    throw logErrorAndStop(
      text || `Expected value '${value}' to be 'string' but was '${typeOf(value)}'.`
    );
  }
};

export const assertNotEqual = (value1: any, value2: any, text?: string) => {
  if (qDev) {
    if (value1 !== value2) return;
    throw logErrorAndStop(text || `Expected '${value1}' !== '${value2}'.`);
  }
};

export const assertEqual = (value1: any, value2: any, text?: string) => {
  if (qDev) {
    if (value1 === value2) return;
    throw logErrorAndStop(text || `Expected '${value1}' === '${value2}'.`);
  }
};

export const assertLessOrEqual = (value1: any, value2: any, text?: string) => {
  if (qDev) {
    if (value1 <= value2) return;
    throw logErrorAndStop(text || `Expected '${value1}' <= '${value2}'.`);
  }
};

export const assertLess = (value1: any, value2: any, text?: string) => {
  if (qDev) {
    if (value1 < value2) return;
    throw logErrorAndStop(text || `Expected '${value1}' < '${value2}'.`);
  }
};

export const assertGreaterOrEqual = (value1: any, value2: any, text?: string) => {
  if (qDev) {
    if (value1 >= value2) return;
    throw logErrorAndStop(text || `Expected '${value1}' >= '${value2}'.`);
  }
};

export const assertGreater = (value1: any, value2: any, text?: string) => {
  if (qDev) {
    if (value1 > value2) return;
    throw logErrorAndStop(text || `Expected '${value1}' > '${value2}'.`);
  }
};

const typeOf = (value: any) => {
  if (value === null) return 'null';
  const type = typeof value;
  if (type === 'object') {
    return value?.constructor?.name || '<unknown>';
  } else {
    return type;
  }
};
