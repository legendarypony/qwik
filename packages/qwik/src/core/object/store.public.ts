import { getDocument } from '../util/dom';
import { isDocument } from '../util/element';
import { QContainerAttr } from '../util/markers';
import { qDev } from '../util/qdev';
import { snapshotState } from './store';
import type { SnapshotState, SnapshotResult } from './store';

// <docs markdown="../readme.md#pauseContainer">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#pauseContainer instead)
/**
 * Serialize the current state of the application into DOM
 *
 * @alpha
 */
// </docs>
export function pauseContainer(elmOrDoc: Element | Document): SnapshotResult {
  const doc = getDocument(elmOrDoc);
  const containerEl = isDocument(elmOrDoc) ? elmOrDoc.documentElement : elmOrDoc;
  const parentJSON = isDocument(elmOrDoc) ? elmOrDoc.body : containerEl;
  const data = snapshotState(containerEl);
  const script = doc.createElement('script');
  script.setAttribute('type', 'qwik/json');
  script.textContent = JSON.stringify(data.state, undefined, qDev ? '  ' : undefined);
  parentJSON.appendChild(script);
  containerEl.setAttribute(QContainerAttr, 'paused');
  return data;
}

export type { SnapshotState, SnapshotResult };
