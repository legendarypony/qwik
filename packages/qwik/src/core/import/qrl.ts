import { EMPTY_ARRAY } from '../util/flyweight';
import type { QRL } from './qrl.public';
import { assertQrl, createQRL, emitEvent, getSymbolHash, QRLInternal } from './qrl-class';
import { isFunction, isString } from '../util/types';
import { getDocument } from '../util/dom';
import { logError } from '../util/log';
import {
  codeToText,
  qError,
  QError_dynamicImportFailed,
  QError_runtimeQrlNoElement,
  QError_unknownTypeArgument,
} from '../error/error';
import { qDev, qTest } from '../util/qdev';
import { getPlatform } from '../platform/platform';
import type { QwikElement } from '../render/dom/virtual-element';
import type { QContext } from '../props/props';
import { assertTrue } from '../assert/assert';
import { isElement } from '../../testing/html';

let runtimeSymbolId = 0;
const RUNTIME_QRL = '/runtimeQRL';
const INLINED_QRL = '/inlinedQRL';

// https://regexr.com/68v72
const EXTRACT_IMPORT_PATH = /\(\s*(['"])([^\1]+)\1\s*\)/;

// https://regexr.com/690ds
const EXTRACT_SELF_IMPORT = /Promise\s*\.\s*resolve/;

// https://regexr.com/6a83h
const EXTRACT_FILE_NAME = /[\\/(]([\w\d.\-_]+\.(js|ts)x?):/;

const QRLcache = new Map<string, string>();

// <docs markdown="../readme.md#qrl">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#qrl instead)
/**
 * Used by Qwik Optimizer to point to lazy-loaded resources.
 *
 * This function should be used by the Qwik Optimizer only. The function should not be directly
 * referred to in the source code of the application.
 *
 * @see `QRL`, `$(...)`
 *
 * @param chunkOrFn - Chunk name (or function which is stringified to extract chunk name)
 * @param symbol - Symbol to lazy load
 * @param lexicalScopeCapture - a set of lexically scoped variables to capture.
 * @alpha
 */
// </docs>
export const qrl = <T = any>(
  chunkOrFn: string | (() => Promise<any>),
  symbol: string,
  lexicalScopeCapture: any[] = EMPTY_ARRAY
): QRL<T> => {
  let chunk = '';
  let symbolFn: null | (() => Promise<Record<string, any>>) = null;
  if (isString(chunkOrFn)) {
    chunk = chunkOrFn;
  } else if (isFunction(chunkOrFn)) {
    symbolFn = chunkOrFn;
    const cached = QRLcache.get(symbol);
    if (!qDev && cached) {
      chunk = cached;
    } else {
      let match: RegExpMatchArray | null;
      const srcCode = String(chunkOrFn);
      if ((match = srcCode.match(EXTRACT_IMPORT_PATH)) && match[2]) {
        chunk = match[2];
      } else if ((match = srcCode.match(EXTRACT_SELF_IMPORT))) {
        const ref = 'QWIK-SELF';
        const frames = new Error(ref).stack!.split('\n');
        const start = frames.findIndex((f) => f.includes(ref));
        const frame = frames[start + 2];
        match = frame.match(EXTRACT_FILE_NAME);
        if (!match) {
          chunk = 'main';
        } else {
          chunk = match[1];
        }
      } else {
        throw qError(QError_dynamicImportFailed, srcCode);
      }
      if (!qDev) {
        // Add to cache
        QRLcache.set(symbol, chunk);
        // Emit event
        emitEvent('qprefetch', {
          symbols: [getSymbolHash(symbol)],
        });
      }
    }
  } else {
    throw qError(QError_unknownTypeArgument, chunkOrFn);
  }

  // Unwrap subscribers
  return createQRL<T>(chunk, symbol, null, symbolFn, null, lexicalScopeCapture, null);
};

export const runtimeQrl = <T>(
  symbol: T,
  lexicalScopeCapture: any[] = EMPTY_ARRAY
): QRLInternal<T> => {
  return createQRL<T>(
    RUNTIME_QRL,
    's' + runtimeSymbolId++,
    symbol,
    null,
    null,
    lexicalScopeCapture,
    null
  );
};

/**
 * @alpha
 */
export const inlinedQrl = <T>(
  symbol: T,
  symbolName: string,
  lexicalScopeCapture: any[] = EMPTY_ARRAY
): QRL<T> => {
  // Unwrap subscribers
  return createQRL<T>(INLINED_QRL, symbolName, symbol, null, null, lexicalScopeCapture, null);
};

export interface QRLSerializeOptions {
  $element$?: QwikElement;
  $getObjId$?: (obj: any) => string | null;
  $addRefMap$?: (obj: any) => number;
}

export const stringifyQRL = (qrl: QRLInternal, opts: QRLSerializeOptions = {}) => {
  assertQrl(qrl);
  let symbol = qrl.$symbol$;
  let chunk = qrl.$chunk$;
  const refSymbol = qrl.$refSymbol$ ?? symbol;
  const platform = getPlatform();

  if (platform) {
    const result = platform.chunkForSymbol(refSymbol);
    if (result) {
      chunk = result[1];
      if (!qrl.$refSymbol$) {
        symbol = result[0];
      }
    }
  }
  if (chunk.startsWith('./')) {
    chunk = chunk.slice(2);
  }
  const parts: string[] = [chunk];
  if (symbol && symbol !== 'default') {
    if (chunk === RUNTIME_QRL && qTest) {
      symbol = '_';
    }
    parts.push('#', symbol);
  }
  const capture = qrl.$capture$;
  const captureRef = qrl.$captureRef$;
  if (captureRef && captureRef.length) {
    if (opts.$getObjId$) {
      const capture = captureRef.map(opts.$getObjId$);
      parts.push(`[${capture.join(' ')}]`);
    } else if (opts.$addRefMap$) {
      const capture = captureRef.map(opts.$addRefMap$);
      parts.push(`[${capture.join(' ')}]`);
    }
  } else if (capture && capture.length > 0) {
    parts.push(`[${capture.join(' ')}]`);
  }
  return parts.join('');
};

export const serializeQRLs = (existingQRLs: QRLInternal<any>[], elCtx: QContext): string => {
  assertTrue(isElement(elCtx.$element$), 'Element must be an actual element');
  const opts: QRLSerializeOptions = {
    $element$: elCtx.$element$,
    $addRefMap$: (obj) => addToArray(elCtx.$refMap$, obj),
  };
  return existingQRLs.map((qrl) => stringifyQRL(qrl, opts)).join('\n');
};

export const qrlToUrl = (element: Element, qrl: QRLInternal): URL => {
  return new URL(stringifyQRL(qrl), getDocument(element).baseURI);
};

/**
 * `./chunk#symbol[captures]
 */
export const parseQRL = (qrl: string, containerEl?: Element): QRLInternal => {
  const endIdx = qrl.length;
  const hashIdx = indexOf(qrl, 0, '#');
  const captureIdx = indexOf(qrl, hashIdx, '[');

  const chunkEndIdx = Math.min(hashIdx, captureIdx);
  const chunk = qrl.substring(0, chunkEndIdx);

  const symbolStartIdx = hashIdx == endIdx ? hashIdx : hashIdx + 1;
  const symbolEndIdx = captureIdx;
  const symbol =
    symbolStartIdx == symbolEndIdx ? 'default' : qrl.substring(symbolStartIdx, symbolEndIdx);

  const captureStartIdx = captureIdx;
  const captureEndIdx = endIdx;
  const capture =
    captureStartIdx === captureEndIdx
      ? EMPTY_ARRAY
      : qrl.substring(captureStartIdx + 1, captureEndIdx - 1).split(' ');

  if (chunk === RUNTIME_QRL) {
    logError(codeToText(QError_runtimeQrlNoElement), qrl);
  }
  const iQrl = createQRL<any>(chunk, symbol, null, null, capture, null, null);
  if (containerEl) {
    iQrl.$setContainer$(containerEl);
  }
  return iQrl;
};

const indexOf = (text: string, startIdx: number, char: string) => {
  const endIdx = text.length;
  const charIdx = text.indexOf(char, startIdx == endIdx ? 0 : startIdx);
  return charIdx == -1 ? endIdx : charIdx;
};

const addToArray = (array: any[], obj: any) => {
  const index = array.indexOf(obj);
  if (index === -1) {
    array.push(obj);
    return array.length - 1;
  }
  return index;
};
