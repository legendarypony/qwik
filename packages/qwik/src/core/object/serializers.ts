import { Component, componentQrl, isQwikComponent } from '../component/component.public';
import { parseQRL, stringifyQRL } from '../import/qrl';
import { isQrl, QRLInternal } from '../import/qrl-class';
import type { QRL } from '../import/qrl.public';
import type { ContainerState } from '../render/container';
import { isResourceReturn, parseResourceReturn, serializeResource } from '../use/use-resource';
import {
  isSubscriberDescriptor,
  parseWatch,
  ResourceReturn,
  serializeWatch,
  SubscriberDescriptor,
} from '../use/use-watch';
import { isDocument } from '../util/element';
import type { GetObject, GetObjID } from './store';

/**
 * 0, 8, 9, A, B, C, D
\0: null character (U+0000 NULL) (only if the next character is not a decimal digit; else itâ€™s an octal escape sequence)
\b: backspace (U+0008 BACKSPACE)
\t: horizontal tab (U+0009 CHARACTER TABULATION)
\n: line feed (U+000A LINE FEED)
\v: vertical tab (U+000B LINE TABULATION)
\f: form feed (U+000C FORM FEED)
\r: carriage return (U+000D CARRIAGE RETURN)
\": double quote (U+0022 QUOTATION MARK)
\': single quote (U+0027 APOSTROPHE)
\\: backslash (U+005C REVERSE SOLIDUS)
 */
export const UNDEFINED_PREFIX = '\u0001';

export interface Serializer<T> {
  prefix: string;
  /**
   * Return true if this serializer can serialize the given object.
   */
  test: (obj: any) => boolean;
  /**
   * Convert the object to a string.
   */
  serialize: ((obj: T, getObjID: GetObjID, containerState: ContainerState) => string) | undefined;
  /**
   * Deserialize the object.
   */
  prepare: (data: string, containerState: ContainerState, doc: Document) => T;
  /**
   * Second pass to fill in the object.
   */
  fill: ((obj: T, getObject: GetObject, containerState: ContainerState) => void) | undefined;
}

const QRLSerializer: Serializer<QRLInternal> = {
  prefix: '\u0002',
  test: (v) => isQrl(v),
  serialize: (obj, getObjId, containerState) => {
    return stringifyQRL(obj, {
      $getObjId$: getObjId,
    });
  },
  prepare: (data, containerState) => {
    return parseQRL(data, containerState.$containerEl$);
  },
  fill: (qrl, getObject) => {
    if (qrl.$capture$ && qrl.$capture$.length > 0) {
      qrl.$captureRef$ = qrl.$capture$.map(getObject);
      qrl.$capture$ = null;
    }
  },
};

const WatchSerializer: Serializer<SubscriberDescriptor> = {
  prefix: '\u0003',
  test: (v) => isSubscriberDescriptor(v),
  serialize: (obj, getObjId) => serializeWatch(obj, getObjId),
  prepare: (data) => parseWatch(data) as any,
  fill: (watch, getObject) => {
    watch.$el$ = getObject(watch.$el$ as any);
    watch.$qrl$ = getObject(watch.$qrl$ as any);
    if (watch.$resource$) {
      watch.$resource$ = getObject(watch.$resource$ as any);
    }
  },
};

const ResourceSerializer: Serializer<ResourceReturn<any>> = {
  prefix: '\u0004',
  test: (v) => isResourceReturn(v),
  serialize: (obj, getObjId) => {
    return serializeResource(obj, getObjId);
  },
  prepare: (data) => {
    return parseResourceReturn(data);
  },
  fill: (resource, getObject) => {
    if (resource.state === 'resolved') {
      resource.resolved = getObject(resource.resolved);
      resource.promise = Promise.resolve(resource.resolved);
    } else if (resource.state === 'rejected') {
      const p = Promise.reject(resource.error);
      p.catch(() => null);
      resource.error = getObject(resource.error);
      resource.promise = p;
    }
  },
};

const URLSerializer: Serializer<URL> = {
  prefix: '\u0005',
  test: (v) => v instanceof URL,
  serialize: (obj) => obj.href,
  prepare: (data) => new URL(data),
  fill: undefined,
};

const DateSerializer: Serializer<Date> = {
  prefix: '\u0006',
  test: (v) => v instanceof Date,
  serialize: (obj) => obj.toISOString(),
  prepare: (data) => new Date(data),
  fill: undefined,
};

const RegexSerializer: Serializer<RegExp> = {
  prefix: '\u0007',
  test: (v) => v instanceof RegExp,
  serialize: (obj) => `${obj.flags} ${obj.source}`,
  prepare: (data) => {
    const space = data.indexOf(' ');
    const source = data.slice(space + 1);
    const flags = data.slice(0, space);
    return new RegExp(source, flags);
  },
  fill: undefined,
};

const ErrorSerializer: Serializer<Error> = {
  prefix: '\u000E',
  test: (v) => v instanceof Error,
  serialize: (obj) => {
    return obj.message;
  },
  prepare: (text) => {
    const err = new Error(text);
    err.stack = undefined;
    return err;
  },
  fill: undefined,
};

const DocumentSerializer: Serializer<Document> = {
  prefix: '\u000F',
  test: (v) => isDocument(v),
  serialize: undefined,
  prepare: (_, _c, doc) => {
    return doc;
  },
  fill: undefined,
};

export const SERIALIZABLE_STATE = Symbol('serializable-data');
const ComponentSerializer: Serializer<Component<any>> = {
  prefix: '\u0010',
  test: (obj) => isQwikComponent(obj),
  serialize: (obj, getObjId, containerState) => {
    const [qrl]: [QRLInternal] = (obj as any)[SERIALIZABLE_STATE];
    return stringifyQRL(qrl, {
      $getObjId$: getObjId,
    });
  },
  prepare: (data, containerState) => {
    const optionsIndex = data.indexOf('{');
    const qrlString = optionsIndex == -1 ? data : data.slice(0, optionsIndex);
    const qrl: QRL<any> = parseQRL(qrlString, containerState.$containerEl$);
    return componentQrl(qrl);
  },
  fill: (component, getObject) => {
    const [qrl]: [QRLInternal] = (component as any)[SERIALIZABLE_STATE];
    if (qrl.$capture$ && qrl.$capture$.length > 0) {
      qrl.$captureRef$ = qrl.$capture$.map(getObject);
      qrl.$capture$ = null;
    }
  },
};

const PureFunctionSerializer: Serializer<Function> = {
  prefix: '\u0011',
  test: (obj) => typeof obj === 'function' && obj.__qwik_serializable__ !== undefined,
  serialize: (obj) => {
    return obj.toString();
  },
  prepare: (data) => {
    const fn = new Function('return ' + data)();
    fn.__qwik_serializable__ = true;
    return fn;
  },
  fill: undefined,
};

const serializers: Serializer<any>[] = [
  QRLSerializer,
  WatchSerializer,
  ResourceSerializer,
  URLSerializer,
  DateSerializer,
  RegexSerializer,
  ErrorSerializer,
  DocumentSerializer,
  ComponentSerializer,
  PureFunctionSerializer,
];

export const canSerialize = (obj: any): boolean => {
  for (const s of serializers) {
    if (s.test(obj)) {
      return true;
    }
  }
  return false;
};

export const serializeValue = (obj: any, getObjID: GetObjID, containerState: ContainerState) => {
  for (const s of serializers) {
    if (s.test(obj)) {
      let value = s.prefix;
      if (s.serialize) {
        value += s.serialize(obj, getObjID, containerState);
      }
      return value;
    }
  }
  return undefined;
};

export interface Parser {
  prepare(data: string): any;
  fill(obj: any): boolean;
}

export const createParser = (
  getObject: GetObject,
  containerState: ContainerState,
  doc: Document
): Parser => {
  const map = new Map<any, Serializer<any>>();
  return {
    prepare(data: string) {
      for (const s of serializers) {
        const prefix = s.prefix;
        if (data.startsWith(prefix)) {
          const value = s.prepare(data.slice(prefix.length), containerState, doc);
          if (s.fill) {
            map.set(value, s);
          }
          return value;
        }
      }
      return data;
    },
    fill(obj: any) {
      const serializer = map.get(obj);
      if (serializer) {
        serializer.fill!(obj, getObject, containerState);
        return true;
      }
      return false;
    },
  };
};
