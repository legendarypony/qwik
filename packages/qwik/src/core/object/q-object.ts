import { assertEqual, assertTrue } from '../assert/assert';
import { qError, QError_immutableProps, QError_verifySerializable } from '../error/error';
import { isQrl } from '../import/qrl-class';
import { tryGetInvokeContext } from '../use/use-core';
import { isDocument, isNode, isQwikElement } from '../util/element';
import { logWarn } from '../util/log';
import { qDev } from '../util/qdev';
import { tryGetContext } from '../props/props';
import { RenderEvent } from '../util/markers';
import { isArray, isFunction, isObject, isSerializableObject } from '../util/types';
import { isPromise } from '../util/promises';
import { canSerialize } from './serializers';
import type { ContainerState, LocalSubscriptionManager, SubscriberMap } from '../render/container';
import type { Subscriber } from '../use/use-watch';

export type QObject<T extends {}> = T & { __brand__: 'QObject' };

export const QObjectRecursive = 1 << 0;
export const QObjectImmutable = 1 << 1;

/**
 * @internal
 */
export const _IMMUTABLE = Symbol('IMMUTABLE');

/**
 * Creates a proxy that notifies of any writes.
 */
export const getOrCreateProxy = <T extends object>(
  target: T,
  containerState: ContainerState,
  flags = 0
): T => {
  const proxy = containerState.$proxyMap$.get(target);
  if (proxy) {
    return proxy;
  }
  return createProxy(target, containerState, flags, undefined);
};

export const createProxy = <T extends object>(
  target: T,
  containerState: ContainerState,
  flags: number,
  subs?: Map<Element, Set<string>>
): T => {
  assertEqual(unwrapProxy(target), target, 'Unexpected proxy at this location', target);
  assertTrue(!containerState.$proxyMap$.has(target), 'Proxy was already created', target);
  assertTrue(isObject(target), 'Target must be an object');
  assertTrue(
    isSerializableObject(target) || isArray(target),
    'Target must be a serializable object'
  );

  const manager = containerState.$subsManager$.$getLocal$(target, subs);
  const proxy = new Proxy(
    target,
    new ReadWriteProxyHandler(containerState, manager, flags)
  ) as any as T;
  containerState.$proxyMap$.set(target, proxy);
  return proxy;
};

const QOjectTargetSymbol = Symbol();
const QOjectSubsSymbol = Symbol();
const QOjectFlagsSymbol = Symbol();

export type TargetType = Record<string | symbol, any>;

class ReadWriteProxyHandler implements ProxyHandler<TargetType> {
  constructor(
    private $containerState$: ContainerState,
    private $manager$: LocalSubscriptionManager,
    private $flags$: number
  ) {}

  get(target: TargetType, prop: string | symbol): any {
    if (typeof prop === 'symbol') {
      if (prop === QOjectTargetSymbol) return target;
      if (prop === QOjectFlagsSymbol) return this.$flags$;
      if (prop === QOjectSubsSymbol) return this.$manager$.$subs$;
      return target[prop];
    }
    let subscriber: Subscriber | undefined | null;
    const invokeCtx = tryGetInvokeContext();
    const recursive = (this.$flags$ & QObjectRecursive) !== 0;
    const immutable = (this.$flags$ & QObjectImmutable) !== 0;
    const value = target[prop];
    if (invokeCtx) {
      subscriber = invokeCtx.$subscriber$;
    }
    if (immutable) {
      // If property is not declared in the target
      // or the prop is immutable, then we dont need to subscribe
      if (!(prop in target) || target[_IMMUTABLE]?.includes(prop)) {
        subscriber = null;
      }
    }
    if (subscriber) {
      const isA = isArray(target);
      this.$manager$.$addSub$(subscriber, isA ? undefined : prop);
    }
    return recursive ? wrap(value, this.$containerState$) : value;
  }

  set(target: TargetType, prop: string | symbol, newValue: any): boolean {
    if (typeof prop === 'symbol') {
      target[prop] = newValue;
      return true;
    }
    const immutable = (this.$flags$ & QObjectImmutable) !== 0;
    if (immutable) {
      throw qError(QError_immutableProps);
    }
    const recursive = (this.$flags$ & QObjectRecursive) !== 0;
    const unwrappedNewValue = recursive ? unwrapProxy(newValue) : newValue;
    if (qDev) {
      verifySerializable(unwrappedNewValue);
      const invokeCtx = tryGetInvokeContext();
      if (invokeCtx && invokeCtx.$event$ === RenderEvent) {
        logWarn(
          'State mutation inside render function. Move mutation to useWatch(), useClientEffect() or useServerMount()',
          invokeCtx.$hostElement$,
          prop
        );
      }
    }
    const isA = isArray(target);
    if (isA) {
      target[prop as any] = unwrappedNewValue;
      this.$manager$.$notifySubs$();
      return true;
    }

    const oldValue = target[prop];
    if (oldValue !== unwrappedNewValue) {
      target[prop] = unwrappedNewValue;
      this.$manager$.$notifySubs$(prop);
    }
    return true;
  }

  has(target: TargetType, property: string | symbol) {
    if (property === QOjectTargetSymbol) return true;
    if (property === QOjectFlagsSymbol) return true;

    return Object.prototype.hasOwnProperty.call(target, property);
  }

  ownKeys(target: TargetType): ArrayLike<string | symbol> {
    let subscriber: Subscriber | null | undefined = null;
    const invokeCtx = tryGetInvokeContext();
    if (invokeCtx) {
      subscriber = invokeCtx.$subscriber$;
    }
    if (subscriber) {
      this.$manager$.$addSub$(subscriber);
    }
    return Object.getOwnPropertyNames(target);
  }
}

const wrap = <T>(value: T, containerState: ContainerState): T => {
  if (isQrl(value)) {
    return value;
  }
  if (isObject(value)) {
    if (Object.isFrozen(value)) {
      return value;
    }
    const nakedValue = unwrapProxy(value);
    if (nakedValue !== value) {
      // already a proxy return;
      return value;
    }
    if (isNode(nakedValue)) {
      return value;
    }
    if (!shouldSerialize(nakedValue)) {
      return value;
    }
    if (qDev) {
      verifySerializable<T>(value);
    }
    const proxy = containerState.$proxyMap$.get(value);
    return proxy ? proxy : getOrCreateProxy(value as any, containerState, QObjectRecursive);
  } else {
    return value;
  }
};

export const verifySerializable = <T>(value: T): T => {
  const seen = new Set();
  return _verifySerializable(value, seen);
};

const _verifySerializable = <T>(value: T, seen: Set<any>): T => {
  const unwrapped = unwrapProxy(value);
  if (unwrapped == null) {
    return value;
  }
  if (shouldSerialize(unwrapped)) {
    if (seen.has(unwrapped)) {
      return value;
    }
    seen.add(unwrapped);
    if (canSerialize(unwrapped)) {
      return value;
    }
    switch (typeof unwrapped) {
      case 'object':
        if (isPromise(unwrapped)) return value;
        if (isQwikElement(unwrapped)) return value;
        if (isDocument(unwrapped)) return value;
        if (isArray(unwrapped)) {
          for (const item of unwrapped) {
            _verifySerializable(item, seen);
          }
          return value;
        }
        if (isSerializableObject(unwrapped)) {
          for (const item of Object.values(unwrapped)) {
            _verifySerializable(item, seen);
          }
          return value;
        }
        break;
      case 'boolean':
      case 'string':
      case 'number':
        return value;
    }
    throw qError(QError_verifySerializable, unwrapped);
  }
  return value;
};
const noSerializeSet = /*#__PURE__*/ new WeakSet<any>();

export const shouldSerialize = (obj: any): boolean => {
  if (isObject(obj) || isFunction(obj)) {
    return !noSerializeSet.has(obj);
  }
  return true;
};

export const fastShouldSerialize = (obj: any): boolean => {
  return !noSerializeSet.has(obj);
};

/**
 * Returned type of the `noSerialize()` function. It will be TYPE or undefined.
 *
 * @see noSerialize
 * @public
 */
export type NoSerialize<T> = (T & { __no_serialize__: true }) | undefined;

// <docs markdown="../readme.md#noSerialize">
// !!DO NOT EDIT THIS COMMENT DIRECTLY!!!
// (edit ../readme.md#noSerialize instead)
/**
 * Marks a property on a store as non-serializable.
 *
 * At times it is necessary to store values on a store that are non-serializable. Normally this
 * is a runtime error as Store wants to eagerly report when a non-serializable property is
 * assigned to it.
 *
 * You can use `noSerialize()` to mark a value as non-serializable. The value is persisted in the
 * Store but does not survive serialization. The implication is that when your application is
 * resumed, the value of this object will be `undefined`. You will be responsible for recovering
 * from this.
 *
 * See: [noSerialize Tutorial](http://qwik.builder.io/tutorial/store/no-serialize)
 *
 * @public
 */
// </docs>
export const noSerialize = <T extends object | undefined>(input: T): NoSerialize<T> => {
  if (input != null) {
    noSerializeSet.add(input);
  }
  return input as any;
};

/**
 * @alpha
 * @deprecated Remove it, not needed anymore
 */
export const mutable = <T>(v: T): T => {
  console.warn(
    'mutable() is deprecated, you can safely remove all usages of mutable() in your code'
  );
  return v;
};

export const isConnected = (sub: Subscriber): boolean => {
  if (isQwikElement(sub)) {
    return !!tryGetContext(sub) || sub.isConnected;
  } else {
    return isConnected(sub.$el$);
  }
};

/**
 * @alpha
 */
export const unwrapProxy = <T>(proxy: T): T => {
  return isObject(proxy) ? getProxyTarget<any>(proxy) ?? proxy : proxy;
};

export const getProxyTarget = <T extends Record<string, any>>(obj: T): T | undefined => {
  return (obj as any)[QOjectTargetSymbol];
};

export const getProxySubs = (obj: any): SubscriberMap | undefined => {
  return (obj as any)[QOjectSubsSymbol];
};

export const getProxyFlags = <T = Record<string, any>>(obj: T): number | undefined => {
  if (isObject(obj)) {
    return (obj as any)[QOjectFlagsSymbol];
  }
  return undefined;
};
