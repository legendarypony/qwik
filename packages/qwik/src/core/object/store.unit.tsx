import { createDocument } from '../../testing/document';
import { useStore } from '../use/use-store.public';
import { getQwikJSON } from './store';
import { useLexicalScope } from '../use/use-lexical-scope.public';
import { component$ } from '../component/component.public';
import { noSerialize } from './q-object';
import { $ } from '../import/qrl.public';
import { logDebug } from '../util/log';
import { runtimeQrl } from '../import/qrl';
import { pauseContainer } from '../object/store';
import { useDocument } from '../use/use-document.public';
import { useHostElement } from '../use/use-host-element.public';
import { suite } from 'uvu';
import { equal } from 'uvu/assert';
import { render } from '../render/dom/render.public';

const storeSuite = suite('store');

storeSuite('should serialize content', async () => {
  const document = createDocument();
  const div = document.createElement('div');
  document.body.appendChild(div);

  await render(
    document.body,
    <div>
      <LexicalScope />
    </div>
  );
  await pauseContainer(document.body);
  const script = getQwikJSON(document.body)!;

  equal(JSON.parse(script.textContent!), {
    ctx: {
      '#1': {
        r: '0 1 2 l 8 e 7 6 h! l j #0 k',
      },
    },
    objs: [
      1,
      'hola',
      {
        a: '3',
        b: '1',
        c: '5',
        d: '6',
        e: '7',
        f: '8',
        g: 'l',
        h: '9',
      },
      {
        thing: '4',
      },
      12,
      123,
      false,
      true,
      null,
      ['0', 'a', '6', 'b', 'c'],
      'string',
      {
        hola: '0',
      },
      ['d'],
      'hello',
      ['0', 'f', '1', 'g'],
      2,
      {},
      {
        count: 'i',
      },
      0,
      '\u0001/runtimeQRL#_',
      '\u0002',
      '\u0000',
    ],
    subs: [],
  });
});

export const LexicalScope_render = () => {
  const [a, b, c, d, e, f, g, h, state, noserialize] = useLexicalScope();
  return (
    <section>
      <p>{JSON.stringify(a)}</p>
      <p>{JSON.stringify(b)}</p>
      <p>{JSON.stringify(c)}</p>
      <p>{String(d)}</p>
      <p>{String(e)}</p>
      <p>{JSON.stringify(f)}</p>
      <p>{JSON.stringify(g)}</p>
      <p>{JSON.stringify(h)}</p>
      <p>{noserialize.text}</p>
      <button onDocumentClick$={() => state.count++}>Rerender {state.count}</button>
    </section>
  );
};

export const LexicalScope = component$(() => {
  const state = useStore({
    count: 0,
  });
  const nu = 1;
  const str = 'hola';
  const obj = {
    a: { thing: 12 },
    b: 'hola',
    c: 123,
    d: false,
    e: true,
    f: null,
    g: undefined,
    h: [1, 'string', false, { hola: 1 }, ['hello']],
  };
  const noserialize = noSerialize({ text: 'not included', window: () => {} });
  const undef = undefined;
  const nulll = null;
  const array = [1, 2, 'hola', {}];
  const boolTrue = true;
  const boolFalse = false;
  const qrl = $(() => logDebug('qrl'));
  const el = useHostElement();
  const doc = useDocument();
  const thing = runtimeQrl(LexicalScope_render, [
    nu,
    str,
    obj,
    undef,
    nulll,
    array,
    boolTrue,
    boolFalse,
    state,
    noserialize,
    qrl,
    el,
    doc,
  ]);
  return <div onClick$={thing}></div>;
});

storeSuite.run();
