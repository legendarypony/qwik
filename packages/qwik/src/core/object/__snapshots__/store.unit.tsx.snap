// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should serialize content 1`] = `
Object {
  "objs": Array [
    1,
    "hola",
    Object {
      "a": "3",
      "b": "1",
      "c": 123,
      "d": "e",
      "e": "d",
      "f": "a",
      "g": "9",
      "h": "4",
    },
    Object {
      "thing": 12,
    },
    Array [
      "0",
      "5",
      "e",
      "6",
      "7",
    ],
    "string",
    Object {
      "hola": "0",
    },
    Array [
      "8",
    ],
    "hello",
    "",
    null,
    Array [
      "0",
      2,
      "1",
      "c",
    ],
    Object {},
    true,
    false,
    Object {
      "count": 0,
    },
    "/runtimeQRL#s1",
    Object {},
    "/runtimeQRL#s0",
    "",
  ],
  "subs": Array [],
}
`;
